File: tests/end-to-end/config/typescript/stack.ts
import * as cdk from 'aws-cdk-lib';
import * as config from 'aws-cdk-lib/aws-config';
import * as ec2 from 'aws-cdk-lib/aws-ec2';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import * as s3 from 'aws-cdk-lib/aws-s3';
import * as sns from 'aws-cdk-lib/aws-sns';

export interface ConfigStackProps extends cdk.StackProps {
  /**
   * @default 'false'
   */
  readonly ec2VolumeAutoEnableIo?: boolean;
  /**
   * @default 'CostCenter'
   */
  readonly ec2VolumeTagKey?: string;
}

/**
 * AWS CloudFormation Sample Template Config: This template demonstrates the usage of AWS Config resources.  **WARNING** You will be billed for the AWS resources used if you create a stack from this template.
 */
export class ConfigStack extends cdk.Stack {
  public readonly configRuleForVolumeTagsArn;
  public readonly configRuleForVolumeTagsConfigRuleId;
  public readonly configRuleForVolumeAutoEnableIoComplianceType;

  public constructor(scope: cdk.App, id: string, props: ConfigStackProps = {}) {
    super(scope, id, props);

    // Applying default props
    props = {
      ...props,
      ec2VolumeAutoEnableIo: props.ec2VolumeAutoEnableIo ?? false,
      ec2VolumeTagKey: props.ec2VolumeTagKey ?? 'CostCenter',
    };

    // Resources
    const configBucket = new s3.CfnBucket(this, 'ConfigBucket', {
    });

    const configTopic = new sns.CfnTopic(this, 'ConfigTopic', {
    });

    const ec2Volume = new ec2.CfnVolume(this, 'Ec2Volume', {
      autoEnableIo: props.ec2VolumeAutoEnableIo!,
      size: 5,
      availabilityZone: cdk.Fn.select(0, cdk.Fn.getAzs('')),
      tags: [
        {
          key: props.ec2VolumeTagKey!,
          value: 'Ec2VolumeTagValue',
        },
      ],
    });

    const lambdaExecutionRole = new iam.CfnRole(this, 'LambdaExecutionRole', {
      assumeRolePolicyDocument: {
        Version: '2012-10-17',
        Statement: [
          {
            Effect: 'Allow',
            Principal: {
              Service: [
                'lambda.amazonaws.com',
              ],
            },
            Action: [
              'sts:AssumeRole',
            ],
          },
        ],
      },
      policies: [
        {
          policyName: 'root',
          policyDocument: {
            Version: '2012-10-17',
            Statement: [
              {
                Effect: 'Allow',
                Action: [
                  'logs:*',
                  'config:PutEvaluations',
                  'ec2:DescribeVolumeAttribute',
                ],
                Resource: '*',
              },
            ],
          },
        },
      ],
    });

    if (configBucket == null) { throw new Error(`A combination of conditions caused 'configBucket' to be undefined. Fixit.`); }
    const configRole = new iam.CfnRole(this, 'ConfigRole', {
      assumeRolePolicyDocument: {
        Version: '2012-10-17',
        Statement: [
          {
            Effect: 'Allow',
            Principal: {
              Service: [
                'config.amazonaws.com',
              ],
            },
            Action: [
              'sts:AssumeRole',
            ],
          },
        ],
      },
      managedPolicyArns: [
        'arn:aws:iam::aws:policy/service-role/AWS_ConfigRole',
      ],
      policies: [
        {
          policyName: 'root',
          policyDocument: {
            Version: '2012-10-17',
            Statement: [
              {
                Effect: 'Allow',
                Action: 's3:GetBucketAcl',
                Resource: [
                  'arn:aws:s3:::',
                  configBucket.ref,
                ].join(''),
              },
              {
                Effect: 'Allow',
                Action: 's3:PutObject',
                Resource: [
                  'arn:aws:s3:::',
                  configBucket.ref,
                  '/AWSLogs/',
                  this.account,
                  '/*',
                ].join(''),
                Condition: {
                  StringEquals: {
                    's3:x-amz-acl': 'bucket-owner-full-control',
                  },
                },
              },
              {
                Effect: 'Allow',
                Action: 'config:Put*',
                Resource: '*',
              },
            ],
          },
        },
      ],
    });

    if (configTopic == null) { throw new Error(`A combination of conditions caused 'configTopic' to be undefined. Fixit.`); }
    const configTopicPolicy = new sns.CfnTopicPolicy(this, 'ConfigTopicPolicy', {
      policyDocument: {
        Id: 'ConfigTopicPolicy',
        Version: '2012-10-17',
        Statement: [
          {
            Effect: 'Allow',
            Principal: {
              Service: 'config.amazonaws.com',
            },
            Action: 'SNS:Publish',
            Resource: '*',
          },
        ],
      },
      topics: [
        configTopic.ref,
      ],
    });

    if (configBucket == null) { throw new Error(`A combination of conditions caused 'configBucket' to be undefined. Fixit.`); }
    if (configTopic == null) { throw new Error(`A combination of conditions caused 'configTopic' to be undefined. Fixit.`); }
    const deliveryChannel = new config.CfnDeliveryChannel(this, 'DeliveryChannel', {
      configSnapshotDeliveryProperties: {
        deliveryFrequency: 'Six_Hours',
      },
      s3BucketName: configBucket.ref,
      snsTopicArn: configTopic.ref,
    });

    if (lambdaExecutionRole == null) { throw new Error(`A combination of conditions caused 'lambdaExecutionRole' to be undefined. Fixit.`); }
    const volumeAutoEnableIoComplianceCheck = new lambda.CfnFunction(this, 'VolumeAutoEnableIOComplianceCheck', {
      code: {
        zipFile: [
          'var aws  = require(\'aws-sdk\');',
          'var config = new aws.ConfigService();',
          'var ec2 = new aws.EC2();',
          'exports.handler = function(event, context) {',
          '    compliance = evaluateCompliance(event, function(compliance, event) {',
          '        var configurationItem = JSON.parse(event.invokingEvent).configurationItem;',
          '        var putEvaluationsRequest = {',
          '            Evaluations: [{',
          '                ComplianceResourceType: configurationItem.resourceType,',
          '                ComplianceResourceId: configurationItem.resourceId,',
          '                ComplianceType: compliance,',
          '                OrderingTimestamp: configurationItem.configurationItemCaptureTime',
          '            }],',
          '            ResultToken: event.resultToken',
          '        };',
          '        config.putEvaluations(putEvaluationsRequest, function(err, data) {',
          '            if (err) context.fail(err);',
          '            else context.succeed(data);',
          '        });',
          '    });',
          '};',
          'function evaluateCompliance(event, doReturn) {',
          '    var configurationItem = JSON.parse(event.invokingEvent).configurationItem;',
          '    var status = configurationItem.configurationItemStatus;',
          '    if (configurationItem.resourceType !== \'AWS::EC2::Volume\' || event.eventLeftScope || (status !== \'OK\' && status !== \'ResourceDiscovered\'))',
          '        doReturn(\'NOT_APPLICABLE\', event);',
          '    else ec2.describeVolumeAttribute({VolumeId: configurationItem.resourceId, Attribute: \'autoEnableIO\'}, function(err, data) {',
          '        if (err) context.fail(err);',
          '        else if (data.AutoEnableIO.Value) doReturn(\'COMPLIANT\', event);',
          '        else doReturn(\'NON_COMPLIANT\', event);',
          '    });',
          '}',
        ].join('\n'),
      },
      handler: 'index.handler',
      runtime: 'nodejs18.x',
      timeout: 30,
      role: lambdaExecutionRole.attrArn,
    });

    if (volumeAutoEnableIoComplianceCheck == null) { throw new Error(`A combination of conditions caused 'volumeAutoEnableIoComplianceCheck' to be undefined. Fixit.`); }
    const configPermissionToCallLambda = new lambda.CfnPermission(this, 'ConfigPermissionToCallLambda', {
      functionName: volumeAutoEnableIoComplianceCheck.attrArn,
      action: 'lambda:InvokeFunction',
      principal: 'config.amazonaws.com',
    });

    if (configRole == null) { throw new Error(`A combination of conditions caused 'configRole' to be undefined. Fixit.`); }
    const configRecorder = new config.CfnConfigurationRecorder(this, 'ConfigRecorder', {
      name: 'default',
      recordingGroup: {
        resourceTypes: [
          'AWS::EC2::Volume',
        ],
      },
      roleArn: configRole.attrArn,
    });

    if (configPermissionToCallLambda == null) { throw new Error(`A combination of conditions caused 'configPermissionToCallLambda' to be undefined. Fixit.`); }
    if (configRecorder == null) { throw new Error(`A combination of conditions caused 'configRecorder' to be undefined. Fixit.`); }
    if (ec2Volume == null) { throw new Error(`A combination of conditions caused 'ec2Volume' to be undefined. Fixit.`); }
    if (volumeAutoEnableIoComplianceCheck == null) { throw new Error(`A combination of conditions caused 'volumeAutoEnableIoComplianceCheck' to be undefined. Fixit.`); }
    const configRuleForVolumeAutoEnableIo = new config.CfnConfigRule(this, 'ConfigRuleForVolumeAutoEnableIO', {
      configRuleName: 'ConfigRuleForVolumeAutoEnableIO',
      scope: {
        complianceResourceId: ec2Volume.ref,
        complianceResourceTypes: [
          'AWS::EC2::Volume',
        ],
      },
      source: {
        owner: 'CUSTOM_LAMBDA',
        sourceDetails: [
          {
            eventSource: 'aws.config',
            messageType: 'ConfigurationItemChangeNotification',
          },
        ],
        sourceIdentifier: volumeAutoEnableIoComplianceCheck.attrArn,
      },
    });
    configRuleForVolumeAutoEnableIo.addDependency(configPermissionToCallLambda);
    configRuleForVolumeAutoEnableIo.addDependency(configRecorder);

    if (configRecorder == null) { throw new Error(`A combination of conditions caused 'configRecorder' to be undefined. Fixit.`); }
    const configRuleForVolumeTags = new config.CfnConfigRule(this, 'ConfigRuleForVolumeTags', {
      inputParameters: {
        tag1Key: 'CostCenter',
      },
      scope: {
        complianceResourceTypes: [
          'AWS::EC2::Volume',
        ],
      },
      source: {
        owner: 'AWS',
        sourceIdentifier: 'REQUIRED_TAGS',
      },
    });
    configRuleForVolumeTags.addDependency(configRecorder);

    // Outputs
    this.configRuleForVolumeTagsArn = configRuleForVolumeTags.attrArn;
    new cdk.CfnOutput(this, 'CfnOutputConfigRuleForVolumeTagsArn', {
      key: 'ConfigRuleForVolumeTagsArn',
      value: this.configRuleForVolumeTagsArn!.toString(),
    });
    this.configRuleForVolumeTagsConfigRuleId = configRuleForVolumeTags.attrConfigRuleId;
    new cdk.CfnOutput(this, 'CfnOutputConfigRuleForVolumeTagsConfigRuleId', {
      key: 'ConfigRuleForVolumeTagsConfigRuleId',
      value: this.configRuleForVolumeTagsConfigRuleId!.toString(),
    });
    this.configRuleForVolumeAutoEnableIoComplianceType = configRuleForVolumeAutoEnableIo.attrComplianceType;
    new cdk.CfnOutput(this, 'CfnOutputConfigRuleForVolumeAutoEnableIOComplianceType', {
      key: 'ConfigRuleForVolumeAutoEnableIOComplianceType',
      value: this.configRuleForVolumeAutoEnableIoComplianceType!.toString(),
    });
  }
}

File: tests/end-to-end/config/golang/stack.go
package main

import (
	cdk "github.com/aws/aws-cdk-go/awscdk/v2"
	config "github.com/aws/aws-cdk-go/awscdk/v2/awsconfig"
	ec2 "github.com/aws/aws-cdk-go/awscdk/v2/awsec2"
	iam "github.com/aws/aws-cdk-go/awscdk/v2/awsiam"
	lambda "github.com/aws/aws-cdk-go/awscdk/v2/awslambda"
	s3 "github.com/aws/aws-cdk-go/awscdk/v2/awss3"
	sns "github.com/aws/aws-cdk-go/awscdk/v2/awssns"
	"github.com/aws/constructs-go/constructs/v10"
	"github.com/aws/jsii-runtime-go"
)

type ConfigStackProps struct {
	cdk.StackProps
	Ec2VolumeAutoEnableIo interface{/* Boolean */}
	Ec2VolumeTagKey *string
}

/// AWS CloudFormation Sample Template Config: This template demonstrates the usage of AWS Config resources.  **WARNING** You will be billed for the AWS resources used if you create a stack from this template.
type ConfigStack struct {
	cdk.Stack
	ConfigRuleForVolumeTagsArn interface{} // TODO: fix to appropriate type
	ConfigRuleForVolumeTagsConfigRuleId interface{} // TODO: fix to appropriate type
	ConfigRuleForVolumeAutoEnableIoComplianceType interface{} // TODO: fix to appropriate type
}

func NewConfigStack(scope constructs.Construct, id string, props *ConfigStackProps) *ConfigStack {
	var sprops cdk.StackProps
	if props != nil {
	}
	stack := cdk.NewStack(scope, &id, &sprops)

	configBucket := s3.NewCfnBucket(
		stack,
		jsii.String("ConfigBucket"),
		&s3.CfnBucketProps{
		},
	)

	configTopic := sns.NewCfnTopic(
		stack,
		jsii.String("ConfigTopic"),
		&sns.CfnTopicProps{
		},
	)

	ec2Volume := ec2.NewCfnVolume(
		stack,
		jsii.String("Ec2Volume"),
		&ec2.CfnVolumeProps{
			AutoEnableIo: props.Ec2VolumeAutoEnableIo,
			Size: jsii.Number(5),
			AvailabilityZone: cdk.Fn_Select(jsii.Number(0), cdk.Fn_GetAzs(jsii.String(""))),
			Tags: &[]*cdk.CfnTag{
				&cdk.CfnTag{
					Key: props.Ec2VolumeTagKey,
					Value: jsii.String("Ec2VolumeTagValue"),
				},
			},
		},
	)

	lambdaExecutionRole := iam.NewCfnRole(
		stack,
		jsii.String("LambdaExecutionRole"),
		&iam.CfnRoleProps{
			AssumeRolePolicyDocument: map[string]interface{} {
				"Version": jsii.String("2012-10-17"),
				"Statement": &[]interface{}{
					map[string]interface{} {
						"Effect": jsii.String("Allow"),
						"Principal": map[string]interface{} {
							"Service": &[]interface{}{
								jsii.String("lambda.amazonaws.com"),
							},
						},
						"Action": &[]interface{}{
							jsii.String("sts:AssumeRole"),
						},
					},
				},
			},
			Policies: &[]*Policy /* FIXME */{
				&Policy/* FIXME */{
					PolicyName: jsii.String("root"),
					PolicyDocument: map[string]interface{} {
						"Version": jsii.String("2012-10-17"),
						"Statement": &[]interface{}{
							map[string]interface{} {
								"Effect": jsii.String("Allow"),
								"Action": &[]interface{}{
									jsii.String("logs:*"),
									jsii.String("config:PutEvaluations"),
									jsii.String("ec2:DescribeVolumeAttribute"),
								},
								"Resource": jsii.String("*"),
							},
						},
					},
				},
			},
		},
	)

	configRole := iam.NewCfnRole(
		stack,
		jsii.String("ConfigRole"),
		&iam.CfnRoleProps{
			AssumeRolePolicyDocument: map[string]interface{} {
				"Version": jsii.String("2012-10-17"),
				"Statement": &[]interface{}{
					map[string]interface{} {
						"Effect": jsii.String("Allow"),
						"Principal": map[string]interface{} {
							"Service": &[]interface{}{
								jsii.String("config.amazonaws.com"),
							},
						},
						"Action": &[]interface{}{
							jsii.String("sts:AssumeRole"),
						},
					},
				},
			},
			ManagedPolicyArns: &[]*string{
				jsii.String("arn:aws:iam::aws:policy/service-role/AWS_ConfigRole"),
			},
			Policies: &[]*Policy /* FIXME */{
				&Policy/* FIXME */{
					PolicyName: jsii.String("root"),
					PolicyDocument: map[string]interface{} {
						"Version": jsii.String("2012-10-17"),
						"Statement": &[]interface{}{
							map[string]interface{} {
								"Effect": jsii.String("Allow"),
								"Action": jsii.String("s3:GetBucketAcl"),
								"Resource": cdk.Fn_Join(jsii.String(""), &[]*string{
									jsii.String("arn:aws:s3:::"),
									configBucket.Ref(),
								}),
							},
							map[string]interface{} {
								"Effect": jsii.String("Allow"),
								"Action": jsii.String("s3:PutObject"),
								"Resource": cdk.Fn_Join(jsii.String(""), &[]*string{
									jsii.String("arn:aws:s3:::"),
									configBucket.Ref(),
									jsii.String("/AWSLogs/"),
									stack.Account(),
									jsii.String("/*"),
								}),
								"Condition": map[string]interface{} {
									"StringEquals": map[string]interface{} {
										"S3XAmzAcl": jsii.String("bucket-owner-full-control"),
									},
								},
							},
							map[string]interface{} {
								"Effect": jsii.String("Allow"),
								"Action": jsii.String("config:Put*"),
								"Resource": jsii.String("*"),
							},
						},
					},
				},
			},
		},
	)

	sns.NewCfnTopicPolicy(
		stack,
		jsii.String("ConfigTopicPolicy"),
		&sns.CfnTopicPolicyProps{
			PolicyDocument: map[string]interface{} {
				"Id": jsii.String("ConfigTopicPolicy"),
				"Version": jsii.String("2012-10-17"),
				"Statement": &[]interface{}{
					map[string]interface{} {
						"Effect": jsii.String("Allow"),
						"Principal": map[string]interface{} {
							"Service": jsii.String("config.amazonaws.com"),
						},
						"Action": jsii.String("SNS:Publish"),
						"Resource": jsii.String("*"),
					},
				},
			},
			Topics: &[]*string{
				configTopic.Ref(),
			},
		},
	)

	config.NewCfnDeliveryChannel(
		stack,
		jsii.String("DeliveryChannel"),
		&config.CfnDeliveryChannelProps{
			ConfigSnapshotDeliveryProperties: &ConfigSnapshotDeliveryProperties/* FIXME */{
				DeliveryFrequency: jsii.String("Six_Hours"),
			},
			S3BucketName: configBucket.Ref(),
			SnsTopicArn: configTopic.Ref(),
		},
	)

	volumeAutoEnableIoComplianceCheck := lambda.NewCfnFunction(
		stack,
		jsii.String("VolumeAutoEnableIOComplianceCheck"),
		&lambda.CfnFunctionProps{
			Code: &Code/* FIXME */{
				ZipFile: cdk.Fn_Join(jsii.String("\n"), &[]*string{
					jsii.String("var aws  = require('aws-sdk');"),
					jsii.String("var config = new aws.ConfigService();"),
					jsii.String("var ec2 = new aws.EC2();"),
					jsii.String("exports.handler = function(event, context) {"),
					jsii.String("    compliance = evaluateCompliance(event, function(compliance, event) {"),
					jsii.String("        var configurationItem = JSON.parse(event.invokingEvent).configurationItem;"),
					jsii.String("        var putEvaluationsRequest = {"),
					jsii.String("            Evaluations: [{"),
					jsii.String("                ComplianceResourceType: configurationItem.resourceType,"),
					jsii.String("                ComplianceResourceId: configurationItem.resourceId,"),
					jsii.String("                ComplianceType: compliance,"),
					jsii.String("                OrderingTimestamp: configurationItem.configurationItemCaptureTime"),
					jsii.String("            }],"),
					jsii.String("            ResultToken: event.resultToken"),
					jsii.String("        };"),
					jsii.String("        config.putEvaluations(putEvaluationsRequest, function(err, data) {"),
					jsii.String("            if (err) context.fail(err);"),
					jsii.String("            else context.succeed(data);"),
					jsii.String("        });"),
					jsii.String("    });"),
					jsii.String("};"),
					jsii.String("function evaluateCompliance(event, doReturn) {"),
					jsii.String("    var configurationItem = JSON.parse(event.invokingEvent).configurationItem;"),
					jsii.String("    var status = configurationItem.configurationItemStatus;"),
					jsii.String("    if (configurationItem.resourceType !== 'AWS::EC2::Volume' || event.eventLeftScope || (status !== 'OK' && status !== 'ResourceDiscovered'))"),
					jsii.String("        doReturn('NOT_APPLICABLE', event);"),
					jsii.String("    else ec2.describeVolumeAttribute({VolumeId: configurationItem.resourceId, Attribute: 'autoEnableIO'}, function(err, data) {"),
					jsii.String("        if (err) context.fail(err);"),
					jsii.String("        else if (data.AutoEnableIO.Value) doReturn('COMPLIANT', event);"),
					jsii.String("        else doReturn('NON_COMPLIANT', event);"),
					jsii.String("    });"),
					jsii.String("}"),
				}),
			},
			Handler: jsii.String("index.handler"),
			Runtime: jsii.String("nodejs18.x"),
			Timeout: jsii.Number(30),
			Role: lambdaExecutionRole.AttrArn(),
		},
	)

	configPermissionToCallLambda := lambda.NewCfnPermission(
		stack,
		jsii.String("ConfigPermissionToCallLambda"),
		&lambda.CfnPermissionProps{
			FunctionName: volumeAutoEnableIoComplianceCheck.AttrArn(),
			Action: jsii.String("lambda:InvokeFunction"),
			Principal: jsii.String("config.amazonaws.com"),
		},
	)

	configRecorder := config.NewCfnConfigurationRecorder(
		stack,
		jsii.String("ConfigRecorder"),
		&config.CfnConfigurationRecorderProps{
			Name: jsii.String("default"),
			RecordingGroup: &RecordingGroup/* FIXME */{
				ResourceTypes: &[]*string{
					jsii.String("AWS::EC2::Volume"),
				},
			},
			RoleArn: configRole.AttrArn(),
		},
	)

	configRuleForVolumeAutoEnableIo := config.NewCfnConfigRule(
		stack,
		jsii.String("ConfigRuleForVolumeAutoEnableIO"),
		&config.CfnConfigRuleProps{
			ConfigRuleName: jsii.String("ConfigRuleForVolumeAutoEnableIO"),
			Scope: &Scope/* FIXME */{
				ComplianceResourceId: ec2Volume.Ref(),
				ComplianceResourceTypes: &[]*string{
					jsii.String("AWS::EC2::Volume"),
				},
			},
			Source: &Source/* FIXME */{
				Owner: jsii.String("CUSTOM_LAMBDA"),
				SourceDetails: &[]*SourceDetail /* FIXME */{
					&SourceDetail/* FIXME */{
						EventSource: jsii.String("aws.config"),
						MessageType: jsii.String("ConfigurationItemChangeNotification"),
					},
				},
				SourceIdentifier: volumeAutoEnableIoComplianceCheck.AttrArn(),
			},
		},
	)

	configRuleForVolumeTags := config.NewCfnConfigRule(
		stack,
		jsii.String("ConfigRuleForVolumeTags"),
		&config.CfnConfigRuleProps{
			InputParameters: map[string]interface{} {
				"Tag1Key": jsii.String("CostCenter"),
			},
			Scope: &Scope/* FIXME */{
				ComplianceResourceTypes: &[]*string{
					jsii.String("AWS::EC2::Volume"),
				},
			},
			Source: &Source/* FIXME */{
				Owner: jsii.String("AWS"),
				SourceIdentifier: jsii.String("REQUIRED_TAGS"),
			},
		},
	)

	return &ConfigStack{
		Stack: stack,
		ConfigRuleForVolumeTagsArn: configRuleForVolumeTags.AttrArn(),
		ConfigRuleForVolumeTagsConfigRuleId: configRuleForVolumeTags.AttrConfigRuleId(),
		ConfigRuleForVolumeAutoEnableIoComplianceType: configRuleForVolumeAutoEnableIo.AttrComplianceType(),
	}
}

